server/Main.hs

```diff
@@ -229,45 +231,54 @@ app sessions users = do
     when (not . isAuthenticated $ session) (Scotty.raiseStatus HTTP.status401 "Please authenticate first")
     Scotty.json @Text $ "This should only be visible when authenticated"
 
-data RegistrationResult = Success | AlreadyRegistered | AttestationError Error deriving (Eq, Show)
+data RegistrationResult
+  = Success
+  | AlreadyRegistered
+  | AttestationError Error
+  deriving (Eq, Show)
+
+handleError :: Show e => Either e a -> Scotty.ActionM a
+handleError (Left x) = Scotty.raiseStatus HTTP.status400 . fromStrict . pack . show $ x
+handleError (Right x) = pure x
 
-handleRegistration :: TVar Sessions -> TVar Users -> Scotty.ActionM ()
-handleRegistration sessions users = do
+finishRegistration :: TVar Sessions -> TVar Users -> Scotty.ActionM ()
+finishRegistration sessions users = do
   (sessionId, session) <- getSessionScotty sessions
   case session of
-    Registering userId challenge -> verifyRegistration sessionId userId challenge
-    _ -> Scotty.raiseStatus HTTP.status400 "You need to be registering to complete registration"
+    Registering userId challenge ->
+      verifyRegistration sessionId userId challenge
+    _ ->
+      Scotty.raiseStatus
+        HTTP.status400
+        "You need to be registering to complete registration"
   where
+    verifyRegistration :: SessionId -> Fido2.UserId -> Fido2.Challenge -> Scotty.ActionM ()
     verifyRegistration sessionId userId challenge = do
-      credentials@(Fido2.PublicKeyCredential {response}) <- Scotty.jsonData @(Fido2.PublicKeyCredential Fido2.AuthenticatorAttestationResponse)
-      liftIO $ print credentials
-      result <- case (verifyAttestationResponse serverOrigin (Fido2.RpId domain) challenge Fido2.UserVerificationRequired response) of
-        Left e -> pure $ AttestationError e
-        Right creds@(Fido2.AttestedCredentialData {credentialId}) -> liftIO $ processCredentials sessionId userId credentialId creds
-      liftIO $ print result
```

Comment:
I actually added this at the very end before sending out my previous PR.

I think it's useful for understanding/debugging. We might want a better logging mechanism, but since this is just a testserver, `liftIO . print` could be enough?

Reply:
i'll add it again